### DISCUSSION TITLE
How to properly execute an incremental SG pipeline only at build?

### DISCUSSION THREAD

**Discussion by TwentyFourMinutes** [dotnet/roslyn#60505] [Q&A]
I am currently working on [Venflow](https://github.com/TwentyFourMinutes/Venflow/tree/reflow) an ORM I am currently trying to re-write from full on reflection to Source Generators. A fairly large portion of the ORM itself and its emitted sources are not relevant for IntelliSense and therefor also do not need any information about the changes during the "lifespan" of the source generator. However, AFAIK I am still required to build the model required for the emitted sources through the usual provider API's. This makes me think it would make sense to only ever run the whole pipeline on build to improve performance even further as building the whole model is quite expensive in itself and has to recompile on even the slightest changes.

Is there any way I could implement such a behaviour at the moment, or would this require new API's? On another note I might also have a wrong take to this problem, but feel free to guide me in a different direction.

Another possible solution I could think of is only collecting the required SyntaxNodes and symbols (saved in a custom type, to avoid using the ISymbol type directly) and doing all of the business logic in the `RegisterSourceOutput`, however I'd still need some way to only ever emit the sources on build.

**Comment by PathogenDavid** [comment]
Use `RegisterImplementationSourceOutput` instead of `RegisterSourceOutput`. (See [the documentation on incremental generator outputs](https://github.com/dotnet/roslyn/blob/6888cd64fd590ed8a3c66fcda9804ab8d6e08b98/docs/features/incremental-generators.md#outputting-values) for more details.)

**Reply by jasonmalinowski** [reply]
Which as a _slight_ caveat is not actually respected by the IDE yet but this API is in place so we can finish that.

**Reply by TwentyFourMinutes** [reply]
So if I understand the docs correctly, the preceding pipeline that is being used to call `RegisterImplementationSourceOutput` will still be executed 'on updates' even if the IDE may respect the special source output?

**Reply by jasonmalinowski** [reply]
Yeah, it'll run like normal for now.

**Comment by dczuo** [comment]
```csharp
var processName = System.Diagnostics.Process.GetCurrentProcess().ProcessName;
return processName == "csc" || processName == "VBCSCompiler" || processName == "dotnet";
```

**Reply by TwentyFourMinutes** [reply]
The idea is very nice and is probably a hacky workaround until the `RegisterImplementationSourceOutput` method is respected by the IDE.

**Reply by jasonmalinowski** [reply]
Note this will break in various scenarios, like hot reload. Please don't do this.
